turbopack/crates/turbo-tasks-backend/src/kv_backing_storage.rs

```diff
@@ -600,6 +640,55 @@ impl<T: KeyValueDatabase + Send + Sync + 'static> BackingStorageSealed
     }
 }
 
+fn restore_strings<D: KeyValueDatabase>(
+    database: &D,
+    tx: &D::ReadTransaction<'_>,
+    global_ids: &LocalIdToGlobalId,
+) -> Result<LocalIdToRcStr> {
+    let mut map = Vec::with_capacity(global_ids.len());
+
+    for global_id in global_ids.iter() {
+        // First, check the in-memory cache
+        let rcstr = {
+            let cache = STRING_CACHE.lock().unwrap_or_else(PoisonError::into_inner);
+            cache.get(&global_id).cloned()
+        };
+
+        let rcstr = if let Some(rcstr) = rcstr {
+            // Cache hit
+            rcstr
+        } else {
+            // Cache miss, fetch from database
+            let Some(value) = database.get(
+                tx,
+                KeySpace::ReverseStringInternMap,
+                IntKey::new(global_id).as_ref(),
+            )?
+            else {
+                bail!("Unable to find string for {global_id}")
+            };
+
+            let rcstr = unsafe {
+                // Safety: We interned a rust string, so it is valid utf-8
+                RcStr::from(str::from_utf8_unchecked(value.borrow()))
+            };
+
+            // Update both caches
+            {
+                let mut cache = STRING_CACHE.lock().unwrap_or_else(PoisonError::into_inner);
+                cache.insert(global_id, rcstr.clone());
```

Comment:
There is a race condition where multiple DB fetches happen in parallel for the same id.

That's pretty rare, so fine for now, but you could easily be a bit more effient in not creating multiple RcStr for this case with `rcstr = cache.entry(global_id).or_insert(rcstr).clone()`